enhancements/installer/custom-dns.md

```diff
@@ -0,0 +1,574 @@
+"openshift-install generate-custom-dns-config" while providing it with a complete
+install-config. After thsi step, when "create cluster" is called with the
+customDNS feature enabled, the validation step will have to ensure both the LB and
+DNS have been configured.
+
+An advantage to this approach is that at the end of "create cluster" the customer
+can expect to have a fully functioning cluster.
+
+This approach suggests a new way to get to a fully functional cluster in 2 parts.
+Does this constitute an "agent in the middle" install? The cluster install is not
+actively waiting for the DNS config to be completed by the customer. Another
+con is that a customer could run the "generate-custom-dns-config" utility
+but never follow through with the DNS config. This will result in dangling LB config.
+
+#### Option 1A: Generate LB and DNS config during pre-install step
+
+Another variation of this approach would be for this pre-install utility to generate
+LB DNS names but not actually configure the Load Balancers. One of the components
+of the LB DNS Name is the infra-id that is generated by the installer. If this
+pre-install utility is run to generate an infra-id, we should make sure to add it
+to the install-config so that eventual call to "create cluster" can use the same
+infra-id. 
+
+The current understanding is that the LB DNS name follows this pattern:
+<infra-id>-ext-<unknown-id>.elb.<region>.<cluster domanin name>
+
+[TODO: We are still figuring out if the "unknown id" is a randomly generated
+load-balancer-resource-id or a known subnet-id.]
+
+So, if the LB DNS name can be predicted by the installer (with the infra id being
+the only randomly generated value within the Installer), then the actual
+configuration of the LB can be postponed to the "create cluster" step with the
+pre-install step just generating the LB DNS name to be used for the DNS configuration.
+
+#### Option 2: Customer configures DNS as a post-install step
+
+Another approach to solving the ordering problem between the LB and DNS configuration
+is by letting "openshift-install create cluster" configure Load Balancers for the
+API, Ingress (and Internal API) services but not DNS configuration for the API and Ingress
+services. "create cluster" provides as its output, the LB DNS names that can then be
+used by the customer to configure their DNS solution.
+
+An advantage of this approach is that we will not be creating an additional step before
+cluster creation.
+
+[TODO: Figure out answers to these questions.]
+Does this solution circumvent the "agent in the middle" problem? The Internal API Server
+would be functioning and so the components within the cluster can communicate with each
+other. So, when "create cluster" command completes, will there be a functioning cluster
+that would not be reachable externally? Will the cluster install successfully without the
+API Server's DNS entries being configured? How about pulling images during cluster install?
```

Comment:
Having users configure DNS solely as a day 2 operation would not work (the installer needs to talk to the api server). One solution that could potentially work, though, is if IP addresses are used pre-install and then those are switched to the LB dns name afterward.

Reply:
Thanks for answering that question about configuring DNS as a day-2 operation.

Configuring DNS in 2 stages, once as a A record with IP addresses and modifying that to be a CNAME record pointing to the LB DNS name will work. I had prematurely rejected that and did not add it to the enhancement. I should add that and get feedback from PM / Managed services regarding the customer workflow.